### DISCUSSION TITLE
How to read bytes from body

### DISCUSSION THREAD

**Discussion by cemoktra** [tokio-rs/axum#288] [Q&A] [answered]
`req: axum::http::Request<axum::body::Body>`

how can i access the bytes?

**Comment by davidpdrsn** [comment, marked as answer]
You have a few options:

- In axum you can use `Bytes` as an extractor which buffers the request body. It also provides other utilities. See [`FromRequest`](https://docs.rs/axum/0.2.3/axum/extract/trait.FromRequest.html) for everything that can be used as extractors.
- Hyper provides [`to_bytes`](https://docs.rs/hyper/0.14.12/hyper/body/fn.to_bytes.html) and [`aggregate`](https://docs.rs/hyper/0.14.12/hyper/body/fn.aggregate.html) that let you buffer the body into a [`Bytes`](https://docs.rs/bytes/1.1.0/bytes/struct.Bytes.html).
- `axum::body::Body` is a re-export of `hyper::Body`, which implements [`http_body::Body`](https://docs.rs/http-body/0.4.3/http_body/trait.Body.html). That gives low level access to the byte stream.

**Comment by demurgos** [comment]
Axum 0.7 provides [`axum::body::to_bytes`](https://docs.rs/axum/latest/axum/body/fn.to_bytes.html) to convert the body into bytes.
